### ISSUE TITLE
Train PWC-Net with two gpus

### ISSUE THREAD

**Issue by lhao0301** [open-mmlab/mmflow#88]
Thanks for your wonderful code!

I have two gpus and to reproduce the PWC-Net, I only need to set the `samples_per_gpu=4` [HERE](https://github.com/open-mmlab/mmflow/blob/0a9a72421a58b9d7c9666620c3ab20be2f94c186/configs/_base_/datasets/flyingchairs_384x448.py#L72) and keep other settings unchanged.
And then, run 
```bash ./tools/dist_train.sh configs/pwcnet/pwcnet_8x1_slong_flyingchairs_384x448.py 2```.

Are these okay?

**Comment by MeowZheng** [comment]
Yes. This change is right

**Comment by lhao0301** [comment]
Btw, have u trained the pwcnet with max-displacement=2[in the ablation of the official pwc paper] other than 4?

**Comment by MeowZheng** [comment]
No, I haven't. Are there any findings from max-displacement=2?

**Comment by lhao0301** [comment]
I just can't reproduce the official epe when max-displacement=2 in my implementation. Hence, I turn to mmflow to check it again.

In my implementation, max-displacement=4, achieve a little better epe.
![image](https://user-images.githubusercontent.com/10356897/153189415-481b6b62-c3ea-49e8-a08f-f14303c374fa.png)

**Comment by MeowZheng** [comment]
Thanks for your sharing.

For max-displacement=2, can I see how you set correlation cfg in config file?

**Comment by lhao0301** [comment]
I plan to change the three places where are commented.

In `configs/_base_/datasets/flyingchairs_384x448.py`:
```
data = dict(
    train_dataloader=dict(
        # samples_per_gpu=1,
        samples_per_gpu=4,
        workers_per_gpu=2,
        drop_last=True,
        ...
```
And in `configs/_base_/models/pwcnet.py`:
```
decoder=dict(
        type='PWCNetDecoder',
        in_channels=dict(
            # level6=81, level5=213, level4=181, level3=149, level2=117),
            level6=81-81+25, level5=213-81+25, level4=181-81+25, level3=149-81+25, level2=117-81+25)),
        flow_div=20.,
        # corr_cfg=dict(type='Correlation', max_displacement=4, padding=0),
        corr_cfg=dict(type='Correlation', max_displacement=2, padding=0),
        ...
```

**Comment by MeowZheng** [comment]
It looks like everything is OK..

**Comment by lhao0301** [comment]
I just start to run training with max-disp=2 with mmflow. I will check the performance of mmflow with different max-disp after finishing the training.
I also found a big difference between my-implementation and mmflow. I sum the loss across batch rather than mean in each level. When max-disp=4, my-implementations seems to be little higher than the epe in mmflow(as reported) also.  A little confused about that  the difference make little impact on the epe.

**Comment by MeowZheng** [comment]
'sum of loss' might impact training, as it is equivalent to changing the lr.

**Comment by lhao0301** [comment]
Hi, after training the model on flyingchairs, I evaluate it and get the result below. It seems a little lower than the official-pwc especially when max-disp=2 also and are the biases okay?

mmflow-pwc-max-disp=4-flyingchairs:
>> sintel-clean = 3.5075257 (vs 3.33 in official-pwc-paper) / sintel-final = 4.769917 (vs 4.59 in official-pwc-paper)
mmflow-pwc-max-disp=2-flyingchairs:
>> sintel-clean = 3.5748842 (vs 3.30 in official-pwc-paper) / sintel-final = 4.8892856 (vs 4.50 in official-pwc-paper)

Btw, I found that the batch_size=8 and only `left/into_future` are used when finetune on flyingthings3d, which is different from the setting(batch_size=4, `left/right/into_future/into_past`) in the official [pwc-repo](https://github.com/NVlabs/PWC-Net/issues/9#issuecomment-405705811). I wonder that are the difference make little of importance?

**Comment by MeowZheng** [comment]
Yes, the performance of  mmflow-pwc-max-disp=4-flyingchais is a little degraded. Based on our limited knowledge, training on fryingchais is just an intermediate step, and little degraded performance might not be essential for final results, but we still welcome and appreciate your contribution to improving performance.

For `left/into_future`, we follow pwcnet training setting from irr, as the official pwcnet repo didn't release any code for dataset setting and just mention it in issue. For batch_size=8, we did some experiences for batch_size 8 or 4, and just found there wasn't a big difference between batch_size=8 and batch_size=4.

ref:
https://github.com/visinf/irr/blob/4d7f6aa46d6989d7dcf8aa1213fbc64f0058e038/datasets/flyingThings3D.py#L228
https://github.com/visinf/irr/blob/4d7f6aa46d6989d7dcf8aa1213fbc64f0058e038/losses.py#L360
https://github.com/visinf/irr/blob/4d7f6aa46d6989d7dcf8aa1213fbc64f0058e038/datasets/flyingThings3D.py#L177

**Comment by lhao0301** [comment]
Okay, **many thanks for solving the confusion in my mind!**

I just met `NAN` during finetuning mmflow-pwc-max-disp=4 on flyingthings3d_subset (I have tried it twice). Have u met the problem? 
<img width="949" alt="5AB0A8EC-1FC0-4745-9B8A-967A2160BC65" src="https://user-images.githubusercontent.com/10356897/154201936-3754d165-0e00-46bc-a8a8-d85b8b517f67.png">

**Comment by MeowZheng** [comment]
Yes.

There is some dirty data in this dataset, and we excluded it.

ref: https://github.com/open-mmlab/mmflow/blob/6da61bf2040b02946e3bd04c4abddb69f7341f5d/mmflow/datasets/flyingthings3d_subset.py#L10-L83

**Comment by lhao0301** [comment]
Yes, I have saw the `REF`. 
Before the training, I changed nothing but the `sample_per_gpu=4`(because of my two gpu-cards). 
Hence, it's a little strange?  Are there anything I missed?

**Comment by MeowZheng** [comment]
It looks like everything is OK..

**Comment by lhao0301** [comment]
I check all the flo files in flyingthings3d_subset dataset and find that , take `left-into-future` for example, `NAN` files are not consistent with that listed in the  [mmflow/mmflow/datasets/flyingthings3d_subset.py](https://github.com/open-mmlab/mmflow/blob/6da61bf2040b02946e3bd04c4abddb69f7341f5d/mmflow/datasets/flyingthings3d_subset.py#L10-L83).

**Comment by MeowZheng** [comment]
This information is so valuable to us. Would you like to share the nan file list with us?

May you create a pull request to revise the list in flyingthings3d_subset?

**Comment by lhao0301** [comment]
Okay.

**Comment by lhao0301** [comment]
Hi, when I finetune the pwcnet model(pretrained on chairs+things3d_subset) on Sintel + KITTI2015 + HD1K, I get `sintel-clean/final-epe=1.54/1.96`, which is **much lower** than the result reported in this repo. Is this okay?

Minor difference:
1. `sample_per_gpu=2`
2.  configuration (the commented lines) in `flow_loss` due to my omission
```
flow_loss=dict(
            type='MultiLevelEPE',
            p=1,
            # q=0.4,
            eps=0.01,
            reduction='sum',
            # resize_flow='upsample',
            weights={
                'level2': 0.005,
                'level3': 0.01,
                'level4': 0.02,
                'level5': 0.08,
                'level6': 0.32
            }),
```

**Comment by MeowZheng** [comment]
q = 4 is for robust loss function
upsample is for KITTI dataset
below is from PWCNet paper
![image](https://user-images.githubusercontent.com/76149310/155260145-7b80b034-1fb7-4886-ae44-bc808f00c4cf.png)

would you like to create a pr to add pretrained model in cfg?

**Comment by lhao0301** [comment]
Yes, I have seen that.
1. It's confused that better epe can be achieved without following the offical settting. I think it's a little strange?
2. Also `p=0.4` should be used when finetune on flyingthings3d_subset according to the paper?
3. Is the ratio of sintel/hd1k/kiiti within per-batch random?

**Comment by MeowZheng** [comment]
1. 'upsample' is necessary for training with KITTI, if we don't set it, the epe for KITTI will be degraded.
2. No. losses for flythings3d and flyingchairs are same.
3. The ratio is fixed. ref:https://github.com/open-mmlab/mmflow/blob/cd82b73eb9c1fb11fd3ed9d0039b715e7f715378/configs/_base_/datasets/sintel_kitti2015_hd1k_320x768.py#L181

**Comment by MeowZheng** [comment]
would you like to create a pr to add pretrained model in cfg?

**Comment by lhao0301** [comment]
Okay, maybe the model now has a bias to sintel rather than kitti.
Yes, I have added the pr.
